### ISSUE TITLE
Increase parsing speed

### ISSUE THREAD

**Issue by FlorianK13** [OpenEnergyPlatform/open-MaStR#546]
This task contains several steps:

1. Search different ways that might increase parsing speed. Parsing is done right now by the `pandas.read_xml` method [here](https://github.com/OpenEnergyPlatform/open-MaStR/blob/d17a92c17912f7f96814d309859dc495f942c07a/open_mastr/xml_download/utils_write_to_database.py#L158). Several alternatives are:
  * `polars`, `duckdb`, `pyspark` might have xml parsers and might be faster
  * use plain xml parsing from python (without pandas)
  * ...

2. Writing to sqlite database right now is done by `pandas.to_sql` [here](https://github.com/OpenEnergyPlatform/open-MaStR/blob/d17a92c17912f7f96814d309859dc495f942c07a/open_mastr/xml_download/utils_write_to_database.py#L182C5-L182C26). There might be other faster methods depending on step 1.
3. Construct a benchmark in an own repository. Use a benchmark xml file from the Marktstammdatenregister and test different implementations for parsing them.
4. Decide for a best method and implement it in `open-mastr`

**Comment by AlexandraImbrisca** [comment]
Hi! I started working on this task and decided to change the steps slightly:

1. Construct the benchmark
    - Use the Marktstammdatenregister to construct a few datasets of various size - ✅  ([link](https://github.com/AlexandraImbrisca/open-MaStR/tree/develop/benchmark))
    - Create a script to automate the calculation and comparison of the parsing speed between various optimisations - ✅  ([link](https://github.com/AlexandraImbrisca/open-MaStR/pull/6))
2. Explore faster methods of parsing the XML
     - Research the options and implement the changes
     - Run the benchmark and analyse the results
3. Explore faster methods of writing to the sqlite database
     - Research the options and implement the changes
     - Run the benchmark and analyse the results
4. Decide on the best method and add it to this repository

**Comment by FlorianK13** [comment]
I was at DACH Energy Informatics Conference and took two points from there:
- Many researchers use open-mastr
- The feature request I heard most often was the question, if we can decrease the time it needs to download and parse the data
I think people will be really happy if this issue is successful 😃

**Comment by nesnoj** [comment]
> I was at DACH Energy Informatics Conference and took two points from there:
> 
> * Many researchers use open-mastr
> * The feature request I heard most often was the question, if we can decrease the time it needs to download and parse the data
>   I think people will be really happy if this issue is successful 😃

Sounds great. I think we cannot do much about the dl speed but I'm really looking forward to the parsing enhancement @AlexandraImbrisca :smile:

**Comment by nesnoj** [comment]
While reviewing #598 I tried to speedup the `to_sql()` for psql DBs by testing different values for `chunksize` and `method` but could not get it siginificantly faster.

Further ideas:
- Use psycopg2's `execute_values()` instead of pandas' `to_sql()`
- Pandas' `to_sql()` with `method=callable`
- Drop indexes before inserting and recreate after
- Check https://github.com/Blue-Yonder-OSS/pg-bulk-loader (also for parallelization)

**Comment by AlexandraImbrisca** [comment]
Thanks a lot for trying and researching these different ways! I can look into postgre after submitting the parallelization

**Comment by nesnoj** [comment]
Some notes on parallelization:
- Please keep in mind the concepts in different OS ([example](https://pythonforthelab.com/blog/differences-between-multiprocessing-windows-and-linux/)). Running successfully under Linux does not imply the same behavior unter Windows and vice versa (if no other packages are used which will handle this for you).
- I propose to give preference to native python packages instead of adding additional requirements, so let's discuss here after your findings on possible ways before the actual implementation.

**Comment by nesnoj** [comment]
I added a new issue for speeding up DB insertion and mp: #600
